<a href="https://colab.research.google.com/github/jmpie1/coursera_cloud_ng/blob/main/ML_TeamWork.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#@title


%%html
<body>
  <marquee style='color: blue;' SCROLLDELAY=60 truespeed='true'>Machine learning</marquee> 


<marquee style='color: blue;' direction="down" height="100" width="200" bgcolor="white" SCROLLDELAY=500 >Données d'entrainement</marquee>
<marquee style='color:blue;' direction="down" height="100" width="200" bgcolor="white" SCROLLDELAY=500>Données de test</marquee>
<marquee style='color: blue;' direction="down" height="100" width="200" bgcolor="white" SCROLLDELAY=500>Modèle</marquee>
<marquee style='color: blue;' direction="down" height="100" width="200" bgcolor="white" SCROLLDELAY=500>Validation de modèle</marquee>
</body>

# <font color='blue'>**Machine Learning pour les novices** </font> <br>
Comment permettre à quelqu'un qui ne s'y connait pas en statistique de comprendre minimalement le fonctionnement des algorithmes de machine learning? <br>

Comment permettre à un profane (dans le domaine) de comprendre, à haut niveau, une problématique de machine learning? <br>

Comment définir de manière très simple les termes clés qui servent à définir le machine learning? <br>

**Cette présentation se veut une réponse à ces questions.**




## <font color='blue'>Importer les librairies nécessaires</font>


In [ ]:
#@title
import random
import numpy as np
import pandas
from IPython import display
from sklearn import linear_model

# Préparation de données
random.seed(1)
x1=np.round( np.array([random.uniform(5,50)*i for i in range(10,30)]),1)
x2= np.array(range(4,24))
location= pandas.DataFrame({'essence':x1, 'duree':x2, 'total': 3.*x2+.09*x1})
x2=[1,2,3,21,22,23,24]
x1= [10,18,20.5, 102,123,133,151.3]
validation = pandas.DataFrame({'essence':x1, 'duree':x2, 'total': ['?','?','?','?','?','?','?']})



## <font color='blue'>Machine learning</font>
On évite volontairement de donner une définition du machine learning, l'objectif est que le lecteur puisse donner une définition du machine learning après avoir lu ce document.

### **Mots-clés**


*   Données d'entrainement
*   Données de test
*   Modèle 
*   Validation de modèle



> #####   **Données d'entrainement**
Une compagnie de location de voiture loue des voitures à des apprentis conducteurs afin qu'ils puissent s'améliorer avant d'aller subir leur examen de conduite. La compagnie facture les clients selon la quantité d'essence consommée  et le nombre de kilomètres parcourus. La présentation du jeu de données est donnée ci-dessous.





In [ ]:
#@title Visualiser les données d'entrainement.
location

essence  duree    total
0     110.5      4   21.945
1     474.5      5   57.705
2     472.4      6   60.516
3     214.2      7   40.278
4     382.1      8   58.389
5     378.4      9   61.056
6     549.1     10   79.419
7     688.4     11   94.956
8     166.0     12   50.940
9     119.2     13   49.728
10    852.2     14  118.698
11    514.0     15   91.260
12    864.7     16  125.823
13    117.2     17   61.548
14    601.0     18  108.090
15    936.7     19  141.303
16    397.7     20   95.793
17   1283.5     21  178.515
18   1275.8     22  180.822
19    184.9     23   85.641



> <em>Description des données</em>


       > essence: nombre de litres d'essence consommée.

       > duree: nombre d'heures écoulées depuis la location de l'auto.

       > total : le montant de la facture.






> ##### **Donnée de test**


Ce jeu de données contient le nombre de litres d'essence consommée ainsi que le nombre de kilomètres effectués pour de nouveaux clients. Toutefois le montant de leur facture est inconnue. La présentation de ce jeu de données est donnée ci-dessous.

In [ ]:
#@title Visualiser les données de test
validation

essence  duree total
0     10.0      1     ?
1     18.0      2     ?
2     20.5      3     ?
3    102.0     21     ?
4    123.0     22     ?
5    133.0     23     ?
6    151.3     24     ?

> ##### **Modèle** 
En analysant les données d'entrainemment, on peut voir que le motant à payer est calculé de la manière suivante.
$$ total = 0.09 \times essence + 3 \times duree $$
Par exemple le client numéro 5 du jeu de données d'entrainement consomme 378.4 litres d'essence et pacours 9 kilomètres. Le motant de la facture de ce client est 61.056 ($0.09 \times 378.4 + 3 \times 9$).
Nous avons un **modèle** linéaire de la forme suivante.
$$total = a_1 \times essence + a_2 \times duree$$
Si on a les valeurs de $a_1$ et de $a_2$, on peut déterminer la facture de n'importe quel client sachant le nombre de litre d'essence consommé et le nombre de kilomètres parcourus.<br>
<em>L'algorithme de machine learning va déterminer $a_1$ et $a_2$ (donc trouver le modèle) pour pouvoir déterminer la facture de nouveaux clients.</em>

> ##### **Validation de Modèle** <br>
La validation consiste à utiliser un jeu de donnée test, en générale, plus petit que celui d'entrainement pour tester le modèle. On utilise une métrique pour évaluer la performence du modèle (voir la section valider le modèle). 



> ### **Mise en oeuvre de l'algoritme avec  [scikit-learn](https://en.wikipedia.org/wiki/Scikit-learn)**
On va utiliser le machine learning pour déterminer le montant de la facture des clients du jeu de données de test. 


In [ ]:
#@title Implémenter le modèle

# Préparation des données d'entrainement
y = (location['total']).values
X = (location[['essence', 'duree']]).values

# Creation d'un conteneur pouvant contenir le modèle
regr = linear_model.LinearRegression()

# Entrainer le modèle à partir de jeu de données d'entrainement. 
# Cette étape vise à détermner les valeurs de a1 et b2.

regression_linéaire = regr.fit(X,y)

#print('Coefficients: \n', regr.coef_)

Coefficients: 
 [0.09 3.  ]


 <em>Le modèle de machine learning est prêt, il peut être testé sur de nouveaux données.</em>

In [ ]:
#@title Les factures des nouveaux clients selon le modèle
# "Prédire" la facture des nouveaux lients
diabetes_y_pred = regression_linéaire.predict((validation[['essence', 'duree']]).values)
for i in range(7):
    print('Pour la ligne ', i, ' du jeu de données de test, le modèle indique:',np.round(np.array(diabetes_y_pred.data[i]),decimals=3))

Pour la ligne  0  du jeu de données de test, le modèle indique: 3.9
Pour la ligne  1  du jeu de données de test, le modèle indique: 7.62
Pour la ligne  2  du jeu de données de test, le modèle indique: 10.845
Pour la ligne  3  du jeu de données de test, le modèle indique: 72.18
Pour la ligne  4  du jeu de données de test, le modèle indique: 77.07
Pour la ligne  5  du jeu de données de test, le modèle indique: 80.97
Pour la ligne  6  du jeu de données de test, le modèle indique: 85.617


> ### Valider le modèle
Est ce que le modèle est valide? J'imagine que le lecteur ne peut pas répondre, pour le moment, à cette question. Car on n'a pas encore défini notre attente. Pour cela nous allons utiliser une <u>métrique</u> simple, elle compte le nombre de données dont l'estimation est correcte. Puisque notre modèle est déterministe, on s'attend à ce que le taux de succès du modèle soit de 100% pour le jeu de données de test. On laisse au lecteur le soin de valider le modèle!

### **En résumé** <br>
Ce tutoriel se veut une présentée du mode de fonctionnement des algorithmes de machine learning. Le lecteur remarquera qu'il n'était nécéssaire d'utiliser un algorithme de machine learning dans ce cas de figure car un tableur excel ferait bien le tavail. Vous avez raison! Toutefois l'objectif de ce tutoriel est justement de permettre à ceux qui ne s'y connait pas en statistique de comprendre la phylosophie des algorithmes de machine de learning.

### <font color='blue'>Pour aller plus loin</font> <br>

#### **Modèle**

> **Variables** <br>
On a un modèle à trois varaibles:


1.   variables explicatives(ou varaible indépendantes): essence, duree
2.   variable expliqué (ou variable indépendante): total



> **Famille du modèle** <br>
Pour entrainer le modèle, il a fallut qu'on ait la facture de certains clients (voir le jeu de données d'entrainement). Autrement, on aurait pas été capable d'entrainer le modèle sans ces exemples de facture. Un tel modèle est dit supervisé.   

> **Paramètres du modèle ($total = a_1 \times essence + a_2  \times duree$)** <br>
Les paramètres du modèle sont  "$a_1$" et "$a_2$". Les valeurs des paramètres sont calculées lors de l'entrainement du modèle à partir des données d'entrainement. Par exemple, si le prix de l'essence change (donc les données d'entrainement vont changer aussi), on doit entrainer le modèle à nouveau pour déterminer les nouvelles valeurs de ces paramètres. Dans ce cas on dit qu'on met à jour le modèle.

> **Analyse du modèle**<br>
Connaissant le nombre de kilomètres parcourus et le nombre de litres d'essence consommée, on peut déterminer avec certitude le total de la facture. Un tel modèle est dite déterministe. Si on voulait plutôt prédire le prix d'un logement en fonction de sa superficié; on serait incapable de prédire le prix avec certitude même si on connait la superficie du logement(il y a d'autres facteurs non contrôlés). Dans ce cas le modèle associé à une telle problématique est dite stochastique. C'est ce genre de problématique qui nécessite l'utilisation du machine learning.


#### **Lien vers des formations sur le machine learning**
> https://learn.datacamp.com/career-tracks/data-scientist-with-python?version=2

> https://www.coursera.org/professional-certificates/ibm-data-science

> https://www.teluq.ca/site/etudes/clom/initiation-vocabulaire-ia.php